Issue #935 WIP: Implementing token based permits in HalfOpen State

KrnSaurabh opened the pull request:
This PR is still under WIP as part of Issue #935 .

We have replaced existing acquire permission approach based on atomic Integer count to a token based permit mechanism. A `token` contains the timestamp when the token was issued to a permitted request. 

New config parameter has beed added to Circuit Breaker as `timeoutIntervalPerCallInHalfOpenState`, post the timeout duration requests will be considered unresponsive and will be marked as completed with Error.

When a request comes in HalfOpen state, `tryAcquirePermission`/`acquirePermission` returns a valid token to permitted request if `permittedNumberOfCallsInHalfOpenState` has not exceeded and stores the issued token into a HashSet. After completing the request, either successful completion or completion with error or ignoredException, request releases the issued token back and token is deleted from HashSet containing the used tokens marking the request appropriately `onError()` or `onSuccess()`.

When a new request comes in and `permittedNumberOfCallsInHalfOpenState` has exceeded, a `CallNotPermittedException` is thrown and no token is being issued to the request. On a request when `permittedNumberOfCallsInHalfOpenState` exceeds, we iterate through HashSet containing used token and remove all the token having `timeoutIntervalPerCallInHalfOpenState` exceeded, considering them unresponsive requests marking them failure by calling `onError()`.

This PR is WIP. Due to change in contract of `tryAcquirePermission()` and `acquirePermission`, many changes has to propagate down to the calling code keeping the new contract. Before I make those changes, I would like to get the implementation reviewed and agreed.
@RobWin Kindly review and provide your suggestions.

Romeh commented:
@KrnSaurabh @RobWin 

Quickly checked the draft implementation , please find out my remarks for our own discussion and validation : 

I do not like we enforce timeout exception for unresponsive calls in half open state for many reasons : 

-  User may not like u are forcing specific timeout for their calls which may not fit their requirements and will like to take control of it
- If they will take control of it ,they may set it to very big value which will take us again to same situation we have now , tokens with very long timeout -> stuck in same state for long ime till the external service http server timeout the TCP connection if it beyond the configured server connection timeout of any
- creating on error time our exceptions will create error events recording in the circuit breaker which can lead to to false statistics and it can impact the state transition if they do not ignore that exception 

what we can do is to free the bucket hashset based into the waiting time u introduced without failing the request which will allow new requests to go through and it can lead to unblock the wait in half open state if required successful test calls are recorded and let the users handle the long running calls by themselves by using proper timeout management through time limiter for example or other ways of doing it .

What do u think ?

KrnSaurabh commented:
@Romeh we had our initial idea of this implementation very much similar to above.  For ref:
https://github.com/resilience4j/resilience4j/issues/935#issuecomment-611474815

Considering we avoid breaking the API and hide the token validation inside the state management of circuit breaker itself, my only concern here is, If we free up the used token buckets and increase atomic Integer count for the token that has expired `timeoutIntervalPerCallInHalfOpenState` , we end up allowing calls more than `permittedNumberOfCallsInHalfOpenState` . 

Actually with this approach, we will keep on allowing calls in HalfOpen state forever until atleast `minimumNumberOfCalls` in HalfOpen state responds with either success or failure and state change occurs to either closed or open. This is because when we free up token that has has expired `timeoutIntervalPerCallInHalfOpenState`, we have no way to know if the call corresponding to that token was success/failure or unresponsive considering we are not changing the contract. Hence we keep deleting the token from hash and incrementing atomic Integer count infinitely until state change occurs to closed or open.

Now in scenario where most of the calls to the end server results into long running/unresponsive call, we will still stuck in HalfOpen state, but off course we will keep allowing calls until it triggers a state change to open or closed.

 This is offcourse is an improvement since now we are giving a chance to application to recover itself by not blocking calls in case of unresponsive calls in HalfOpen state, but I am not sure if we solve the problem completely. 
Also It is possible that some user might want to consider long running/unresponsive calls as failure.

KrnSaurabh commented:
@Romeh @RobWin  Any conclusion on this or shall I go ahead and make changes as suggested by @Romeh  in the last comment ?

RobWin commented:
I have a look at it on Tuesday.  We have public holiday here

RobWin commented:
We always tried to reduce the amount of objects which are created by a CircuitBeraker per call in order to reduce GC pressure.
That's why we never used `Instant` or `Duration`, but `long` and `TimeUnit`.  We should try to avoid to create an `Optional`,  `Token` and an `Instant` per call now.

I also would like to void breaking the API. But how can we achieve this? We need to know which token was used properly in order to delete it from the HashMap.
@Romeh Do you have an idea?

RobWin commented on resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/Token.java:
```diff
@@ -0,0 +1,15 @@
+import java.time.LocalDateTime;
+
+public class Token {
+    private LocalDateTime timestamp;
```
I think we should use `System.nanoTime()` and  `long` as the creation timestamp in order to avoid too many objects.
It could be created inside of the constructor, or?

RobWin commented on resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java:
```diff
@@ -502,27 +508,27 @@ static CircuitBreaker of(String name,
+     * @return {@code Token} if a permission was acquired and {@code Optional.empty()} otherwise
      */
-    boolean tryAcquirePermission();
+    Optional<Token> tryAcquirePermission();
```
Even if I like `Optional` we should try to avoid it in order to reduce GC pressure.

RobWin commented:
@dlsrb6342 @storozhukBM @hexmind What is your opinion?

dlsrb6342 commented:
I agree with @Romeh (https://github.com/resilience4j/resilience4j/pull/1007#issuecomment-633718997) and @hexmind (https://github.com/resilience4j/resilience4j/issues/935#issuecomment-637702579). 
I think we just check `waitDurationInHalfOpenState` and timeout should be handled by users.

RobWin commented:
https://github.com/resilience4j/resilience4j/issues/935#issuecomment-638039472

KrnSaurabh commented:
@RobWin  @hexmind 
Refactored the code as per the comment https://github.com/resilience4j/resilience4j/issues/935#issuecomment-638904524.

Added an optional configuration parameter `waitDurationInHalfOpenState`  for circuit breaker. If the parameter is not set, circuit breaker would behave same as before, i.e circuit breaker would wait in Half Open state infinitely until long running/unresponsive calls completes with success or failure.

`waitDurationInHalfOpenState` if configured, controls how long the CircuitBreaker should stay in Half Open state before it switches to open.

`waitDurationInHalfOpenState` shall be configured to at least 1 second or else it throws error if it is set to less than 1 sec. This constrained shall be discussed and agreed upon.

RobWin commented:
Thanks, I have a look at it

RobWin commented:
The downside I still see is that, if you set waitDurationInHalfOpenState to a larger value like 5 Minutes and a request is stuck, the CB might be stuck multiple minutes until the timeout duration is elapsed.

But at least it's the most simple and non-intrusive solution.

RobWin commented:
I would be fine with this change. Could you please add some tests?

KrnSaurabh commented:
Sure. I will add unit test cases for same.

KrnSaurabh commented:
> The downside I still see is that, if you set waitDurationInHalfOpenState to a larger value like 5 Minutes and a request is stuck, the CB might be stuck a few minutes until the timeout duration is elapsed.

I totally agree with you @RobWin . Token based approach made much more sense to me but had challenges breaking the api. May be we can plan some long term solution later similar to those lines and can even accommodate api changes.

RobWin commented:
Let's do this change for 1.X.
Maybe we can include some new ideas for a 2.x CircuitBreaker.

RobWin commented:
I will merge it as soon as you have added some tests. 👍

dlsrb6342 commented on resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerStateMachine.java:
```diff
@@ -942,6 +942,13 @@ public CircuitBreakerMetrics getMetrics() {
+            final long waitDurationInHalfOpenState = circuitBreakerConfig.getWaitDurationInHalfOpenState().toMillis();
+            if (waitDurationInHalfOpenState >= 1000) {
+                ScheduledExecutorService scheduledExecutorService = schedulerFactory.getScheduler();
+                scheduledExecutorService
```
I think this `ScheduledFuture` should be canceled if Circuit transit to other states. 
https://github.com/resilience4j/resilience4j/issues/991.

RobWin replied:
Yes

RobWin replied:
Yes, we introduces a preTransitionHook -> https://github.com/resilience4j/resilience4j/commit/378dd2b65889b83cda2a2b3d4203e86604b5044e

KrnSaurabh commented:
Kindly review #1042 . I accidentally deleted the remote branch and I have created a new PR #1042  incorporating all the review comments on this PR. Kindly review the same.
Sorry for the inconvenience caused.
